Refactoring of the B0 shimming

Aurelienpujolm opened the pull request:
## Checklist

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied the relevant labels to this PR
- [x] I've added relevant tests for my contribution
- [x] I've updated the documentation and/or added correct docstrings
- [x] I've assigned a reviewer



## Description
This PR is made to refactor the B0 shimming to make it more readable and efficient. 
For this, I've changed the sequencer files in three different classes to make it easier to understand, and to let futur programmers create new sequencer classes by inheriting the Sequencer class. 
This PR, will be followed by two PR to fix the issue of the multiprocessing and the one about resampling.

Aurelienpujolm committed:
Make the refactoring for the sequencer

Aurelienpujolm committed:
Insert the new sequencer in the B0 shimming procedure

Aurelienpujolm committed:
Edit the test sequencer to the refactoring

Aurelienpujolm committed:
Fix an error to allow user to input mask with the wrongs dimensions

Aurelienpujolm committed:
Removed the changes on the b0shim file about the new way of plotting images

po09i reviewed (commented):
These are really big changes, to facilitate the review, can you go through the changes yourself first. For example, things that I noticed at first glance are:
- Don't import with a "*" to avoid unnecessary imports in the future
- The documentation of arguments has not been updated
- Fix all warnings created by PyCharm

Aurelienpujolm committed:
Update b0shim.py to removed all useless imports

Aurelienpujolm committed:
Fixed the documentation and other minors stuff

Aurelienpujolm committed:
Fixed an indent error

Aurelienpujolm committed:
Fixed an error in plot_currents

Aurelienpujolm committed:
Fixed another indent problem in plot_static_riro

po09i commented on shimmingtoolbox/cli/b0shim.py:
```diff
@@ -7,7 +7,6 @@
 - fieldmap_realtime
 - gradient_realtime
 """
-
```
Add the space back

po09i commented on shimmingtoolbox/cli/b0shim.py:
```diff
@@ -21,12 +20,11 @@
+from shimmingtoolbox.shim.sequencer import extend_fmap_to_kernel_size, parse_slices, new_bounds_from_currents
 from shimmingtoolbox.utils import create_output_dir, set_all_loggers, timeit
 from shimmingtoolbox.shim.shim_utils import phys_to_gradient_cs, phys_to_shim_cs, shim_to_phys_cs
-
```
Add space back

po09i commented on shimmingtoolbox/cli/b0shim.py:
```diff
@@ -259,14 +257,16 @@ def dynamic(fname_fmap, fname_anat, fname_mask_anat, method, opt_criteria, slice
+                              reg_factor=reg_factor,
+                              path_output=path_output)
+    #2) Launch shim sequencer
+    coefs = sequencer.shim_sequencer()
```
Rename the method to just "shim" (shim_sequencer --> shim)

po09i commented on shimmingtoolbox/cli/b0shim.py:
```diff
@@ -671,16 +671,17 @@ def realtime_dynamic(fname_fmap, fname_anat, fname_mask_anat_static, fname_mask_
                                       path_output=path_output)
-
+    # 2) Launch the sequencer
+    out = real_time_pmu_sequencer.shim_realtime_pmu_sequencer()
```
Rename the method to just "shim"

po09i commented on shimmingtoolbox/cli/b0shim.py:
```diff
@@ -1221,7 +1222,6 @@ def max_intensity(fname_input, fname_mask, fname_output, verbose):
 
     logger.info(f"Txt file is located here:\n{fname_output}")
 
-
```
Add space back

po09i commented on shimmingtoolbox/shim/sequencer.py:
```diff
@@ -42,838 +38,1063 @@
-        nii_anat (nibabel.Nifti1Image): Nibabel object containing anatomical data in 3d.
-        nii_mask_anat (nibabel.Nifti1Image): 3D anat mask used for the optimizer to shim in the region of interest.
+class Sequencer(object):
+    def _select_optimizer(self, unshimmed, affine, pmu: PmuResp = None):
```
"\_" designate private methods. Since this is now a class, we should make it available in the object. This would allow to change the optimizer instead of recreating a new object. So remove the "_" and make sure that calling it indeed does correctly change the optimizer.

po09i commented on shimmingtoolbox/shim/sequencer.py:
```diff
@@ -42,838 +38,1063 @@
+        self.bounds = None
+        self.extending = True
+
+    def shim_sequencer(self):
```
I think we should move most of the error checking in the init and have the "shim" part be mostly just that, shimming. You could create a local method that includes most of this error checking. I would think that there is a lot of duplication between dynamic sequencer and the realtime one so most of this error checking could probably move to the base sequencer class.

po09i replied:
Also as mentioned above, rename to shim()

po09i replied:
Consider using @property so you can error check every time an attribute is changed. This would allow to be cleaner rather than have a "one function that checks everything" and if one attribute changes then we have to call the function that checks everything again.

Aurelienpujolm replied:
I separated the two parts of the sequencer. However, I do think that is good to keep in the real time sequencer, the calculation of the parameters for the real time shimming. (Because it's not just checking the dimensions of everything, but also a lot of mathematical operation etc.)
If needed, I can also separate it in a different function

kousu replied:
@Aurelienpujolm asked me for some advice. I have not read all the code in detail, but on this point:

1. `check_inputs` should be deleted.

    My instinct is that checking preconditions  be done _in each method_, near where they are used. Perhaps some precondition checking should happen in `__init__`, but it is a lot easier to avoid bugs if each function that assumes it has a 3d array checks that it has 3d array. It means more code but it's boilerplate code that's easy to write and easy to fix if it's wrong, and it avoid the ambiguity about "when/where is it safe to/must I call this `check_inputs()` function?".

    Also `check_inputs()` isn't just checking inputs, it is editing them too, which is misleading:

   https://github.com/shimming-toolbox/shimming-toolbox/blob/fe37b5d1a725a2deaab7697a275489e16928d267/shimmingtoolbox/shim/sequencer.py#L220-L230

    There should only be one line that says `self.nii_fieldmap = ` and it should be in `__init__`. All the transformations/normalizations needed should be done to that data before it is saved as a member variable.
2. But `property()` isn't a good idea either. [Immutability](https://realpython.com/python-mutable-vs-immutable-types/) is.

    Simply disallow people editing the shim parameters. If they want to change the shim parameters, force them to make a new shim object, which then runs through all the precondition checking in `__init__()` (or maybe in `_check_inputs()`, but again I'd recommend that only if it is marked as a private method so that users aren't confused).

kousu replied:
> I think we should move most of the error checking in the init and have the "shim" part be mostly just that, shimming.

I also think you can go further and stop using a class at all: combine `__init__()` and `shim()` into a single **function** (**not a method**) that does all the work in one step, and returns an object with the results (meaning your `.coefs`, `.slices`, etc). Move all the helper methods like `plot_current` and `eval` to their own top-level functions.

An _object_ is supposed to be a little miniature state machine: some data plus some methods that manipulate that data to keep the different pieces of it consistent. But state machines are notoriously hard to get right; it's so easy to add a little edit to some method that subtly invalidates other pieces of data, but only if that method gets called in a certain order. It's a nightmare to keep track of. Immutability is such a much stronger technique that objects are just unnecessary, in my opinion.

I read the addition of [`dataclass`](https://docs.python.org/3/library/dataclasses.html) as python recognizing that objects are a mistake, and structs + top level functions that know how to interpret them are better.

----

And while avoiding objects, also avoid inheritance. Your life is better without it. It is difficult and unnecessary and has wasted a lot of money in the last 20 years. ["Composition" is better](https://www.digitalocean.com/community/tutorials/composition-vs-inheritance), which simply means: write subroutines. That doesn't mean you can't have polymorphism -- you can design APIs and then write a series of functions that fit them, a set that all have the same input and output types or you can design [protocols](https://peps.python.org/pep-0544/) (what other languages call 'interfaces') -- but you can do it without confusing yourself and future maintainers with trying to use inheritance.

po09i commented on shimmingtoolbox/shim/sequencer.py:
```diff
@@ -42,838 +38,1063 @@
+        logger.debug(f"Saved figure: {fname_figure}")
+
+
+class ShimSequencer(Sequencer):
```
Document the attributes of the class (every time you declare self.xxxxx and that it is not a method)

Aurelienpujolm replied:
I documented the attributes, it should be good now !

po09i reviewed (commented):
I started my review, I think this is already a lot of things to change so I will wait for your next changes before continuing
Some other minor comments:
- Remove unnecessary imports
- Add the type information in the Return. (See google documentation style)

Aurelienpujolm committed:
Refactored sequencer to adress the comments 

There is a lot of modification, but most of them were to make sure that the use can't for example break the program by inputing a riro in a static optmization etc.

Aurelienpujolm committed:
Update the b0shim.py files to match the new sequencer

Aurelienpujolm committed:
Update test_sequencer to match the new sequencer

Aurelienpujolm committed:
Fixed a bug if the fieldmap is 2D

Aurelienpujolm committed:
Fixed the 2D fmap bug

Aurelienpujolm committed:
Add spaces in the code

Aurelienpujolm committed:
Document the attributes of the new classes

Aurelienpujolm committed:
Update the documentation

Aurelienpujolm committed:
Update relations of classes 

Update the Sequencer class, to have better documentation and gestion of self between classes

Aurelienpujolm committed:
Update RealTimeSequencer init

Aurelienpujolm committed:
Refactored sequencer.py to used more function

This commit is make to divide sequencer.py with more functions to make it more readable. It's also making a better init for all of the classes

Aurelienpujolm committed:
Fixed the documentation, and divide eval in more functions

Aurelienpujolm committed:
Fixed the documetation by making tuple

po09i commented on shimmingtoolbox/shim/sequencer.py:
```diff
@@ -23,857 +21,1365 @@
                            artefacts.
+        """
+        super().__init__(slices, mask_dilation_kernel, mask_dilation_kernel_size, reg_factor, path_output)
+        self.nii_fieldmap, self.nii_fieldmap_orig, self.extending = self.get_fieldmap(nii_fieldmap)
```
```suggestion
        self.nii_fieldmap, self.nii_fieldmap_orig, self.fmap_is_extended = self.get_fieldmap(nii_fieldmap)
```

Change `self.extending` to `self.fmap_is_extended` since just extending is not clear

po09i commented on shimmingtoolbox/shim/sequencer.py:
```diff
@@ -23,857 +21,1365 @@
+        coefs = self.optimize(self.nii_mask_anat)
+        return coefs
+
+    # TODO : Maybe put the select_optimizer in the init
```
```suggestion

```

po09i commented on shimmingtoolbox/shim/sequencer.py:
```diff
@@ -23,857 +21,1365 @@
+
+        # global supported_optimizers
+        if self.method in supported_optimizers:
+            if self.method == 'least_squares':
+                optimizer = supported_optimizers[self.method](self.coils, unshimmed, affine, self.opt_criteria,
+                                                              reg_factor=self.reg_factor)
```
This will never get executed because if it's realtime, it will use the realtime optimizer:
```
        if self.method == 'least_squares':
            self.method = 'least_squares_rt'
```

We can remove it

Aurelienpujolm replied:
I'm not sure about that, because in the case of a RT shimming, we are first doing a static optimization and then an optimization with the Riro ? If it's the case then we have to make sure to do both optimization ? (I am not sure about that)

po09i replied:
Good catch

Aurelienpujolm committed:
Fixed the comment from Alex and some other modifications 

For evaluate shimming, this commit make sure that we are not calculating everything, unless we are in debug (to not lose time)
In the different functions used in eval, I switched some lines to faster one, using better vectorization of vector (ie : instead of a for loop using np.sum just do a matrices multiplication) 
Removed most of the for loop to make the code faster ( I did not do that in display_shimmed_results, because it was faster to do a for loop instead of a np.ma.... for both of the full masks)

Aurelienpujolm committed:
Fixed a bug

po09i committed:
Merge branch 'master' into ap/B0-shimming-refactoring

po09i committed:
Code review

Aurelienpujolm committed:
Adressed monday's meeting comments

Aurelienpujolm committed:
add a c in a name of a function

Aurelienpujolm committed:
Make a better documentation 

Fixed some typos, and changed some sentences

po09i committed:
Review of documentation

Aurelienpujolm committed:
Fixed the documentation and adressed 27/04/23 meeting

Aurelienpujolm committed:
Update test_sequencer.py to have better coveralls

Aurelienpujolm committed:
Fixed Test_Sequencer

Aurelienpujolm committed:
Fixed minor bug

Aurelienpujolm committed:
Fixed bug

po09i committed:
Change class object names and french todos

Aurelienpujolm committed:
Adressed TODO

po09i committed:
Moved figure to debug since it is mostly unecessary

Aurelienpujolm committed:
Merge remote-tracking branch 'origin/master' into ap/B0-shimming-refactoring

po09i reviewed (approved):
LGTM! Great job! Can you write a comment in the PR about what is the current time/RAM usage you are getting with the same dataset/options/computer used for previous optimization PRs.

Aurelienpujolm commented:
With this PR, a Mac computer 10.15 with a RAM of 16 Go, and 8 hearts,  and a dataset of : 
coil ( 32 channels) ;
anatomic image ([174,174,50] );
fieldmap([112,112,70]);
mask ([112,112,70]).
The time of the B0 shimming is now around 63 seconds VS 80 before this PR(https://github.com/shimming-toolbox/shimming-toolbox/pull/432) . However, this time will be lowered by some new PR : 
- One that will allow the user to save the fig in a faster way (on the same computer it allows me to save around 8 seconds so far)
- One that will save some time by not resampling two times the same mask in some cases (it will save in my case around 15 seconds) 
- Another one that will reimplement multiprocessing in the optimization to save up some time (8 seconds of save)
- A last one to save up some time in calc_shimmed_anat_orient by not resampling the coils on the mask. (5 seconds of save)
With all of these changes, the B0 shimming could take around 25-30 seconds which will be great, and it could be even lower with better computers)
